### PR TITLE
Bug 1480864 - Refactor HistoryPanel

### PR DESCRIPTION
**Pull request by justindarc**
https://bugzilla.mozilla.org/show_bug.cgi?id=1480864

This PR is based on PR #4113.

### CODE REVIEW

**Comment by justindarc** [comment]
NOTE: Because this is essentially a complete rewrite of the HistoryPanel, it may be easier to review the patched `HistoryPanel.swift` as a whole rather than looking at the diff.

**Comment by justindarc** [comment]
This isn't ready for review yet. I'd like to add infinite scrolling to this first.

**Comment by justindarc** [comment]
I've changed my mind. Let's review this and land it on its own and implement infinite scrolling in another bug.

File: Client/Frontend/Home/HistoryPanel.swift [hunk 1]
```diff
@@ -373,169 +258,156 @@ class HistoryPanel: SiteTableViewController, HomePanel {
...
     }
 
-    func showSyncedTabs() {
-        let nextController = RemoteTabsPanel(profile: profile)
-        nextController.homePanelDelegate = self.homePanelDelegate
-        self.refreshControl?.endRefreshing()
-        self.navigationController?.pushViewController(nextController, animated: true)
+    // MARK: - UITableViewDataSource
+    func numberOfSections(in tableView: UITableView) -> Int {
+        return 5
```
Comment (garvankeeley) [inline Client/Frontend/Home/HistoryPanel.swift]: If using the `enum  Sections` code style, then that enum can have a `static let count = 5`.
`ActivityStreamPanel` line 159 has the example I followed for this style.

File: Client/Frontend/Home/DownloadsPanel.swift [hunk 2]
```diff
@@ -58,6 +58,10 @@ struct DateGroupedTableData<T : Equatable> {
     var lastWeek: [(T, TimeInterval)] = []
     var older: [(T, TimeInterval)] = []
 
+    var isEmpty: Bool {
```
Comment (garvankeeley) [inline Client/Frontend/Home/DownloadsPanel.swift]: Move `DateGroupedTableData` to its own file?

File: Client/Frontend/Home/HistoryPanel.swift [hunk 3]
```diff
@@ -373,169 +258,156 @@ class HistoryPanel: SiteTableViewController, HomePanel {
...
+        guard section > 0 else {
+            return nil
+        }
+
+        // Ensure there are rows in this section.
+        guard groupedSites.numberOfItemsForSection(section - 1) > 0 else {
+            return nil
+        }
+
+        switch section - 1 {
```
Comment (garvankeeley) [inline Client/Frontend/Home/HistoryPanel.swift]: recommend to encapsulate in `enum Section ...` 

```
enum Section: Int {
 case a
 case b 

  var title: String? {
            switch self {
            case .a: return Strings.AAA
            case .b: return Strings.BBB
            }
        }
}

File: Client/Frontend/Home/HistoryPanel.swift [hunk 4]
```diff
@@ -373,169 +258,156 @@ class HistoryPanel: SiteTableViewController, HomePanel {
...
 
     override func tableView(_ tableView: UITableView, viewForHeaderInSection section: Int) -> UIView? {
-        if section == 0 {
+        // First section is for Sync/recently closed and its header has no view.
+        guard section > 0 else {
+            return nil
+        }
+
+        // Ensure there are rows in this section.
+        guard groupedSites.numberOfItemsForSection(section - 1) > 0 else {
```
Comment (garvankeeley) [inline Client/Frontend/Home/HistoryPanel.swift]: maybe this whole function will just be`if Section(section).siteCountFor(groupedSites: groupedSites) < 1 return nil`, or is that too odd?

``` enum Section ...
func siteCountFor(groupedSites: groupedSites): Int {
  // if self is not a site section return zero
}

File: Client/Frontend/Home/HistoryPanel.swift [hunk 5]
```diff
@@ -548,19 +420,59 @@ class HistoryPanel: SiteTableViewController, HomePanel {
...
     }
 
+    // MARK: - Empty State
+    func updateEmptyPanelState() {
+        if groupedSites.isEmpty {
+            if emptyStateOverlayView.superview == nil {
+                tableView.addSubview(emptyStateOverlayView)
+                emptyStateOverlayView.snp.makeConstraints { make -> Void in
+                    make.left.right.bottom.equalTo(self.view)
+                    make.top.equalTo(self.view).offset(100)
```
Comment (garvankeeley) [inline Client/Frontend/Home/HistoryPanel.swift]: Please ensure 100 matches the offset for the other empty panel states

File: Client/Frontend/Home/HistoryPanel.swift [hunk 6]
```diff
@@ -93,70 +84,53 @@ class HistoryPanel: SiteTableViewController, HomePanel {
...
-            break
-        case .DynamicFontChanged:
-            if emptyStateOverlayView.superview != nil {
-                emptyStateOverlayView.removeFromSuperview()
+    override func reloadData() {
+        groupedSites = DateGroupedTableData<Site>()
+
+        fetchData().uponQueue(.main) { result in
+            if let sites = result.successValue {
+                for site in sites {
```
[the remark this answers is not in the document]
Comment (garvankeeley) [reply Client/Frontend/Home/HistoryPanel.swift]: nevermind, i was thinking the closure arg of flatMap was non-optional, and it isn't.

File: Client/Frontend/Home/HistoryPanel.swift (hunk 5 above)
Comment (garvankeeley) [reply Client/Frontend/Home/HistoryPanel.swift]: EDIT: nvm, I am sure you didn't change this code, and it is just moved

**Pull request merged**